### PR TITLE
Added basic_qos functionality

### PR DESCRIPTION
**Pull request by MattBearson**
This part of code supports `basic_qos` functionality with default values and setters if a user needs to change it.

### CODE REVIEW

**Comment by stevenklar** [comment]
Thank you for starting this!

In order to merge this properly for a next release these things at least need to be done:

- [ ] Move this code basic.qos related to the Consumer class and make the methods private or protected, or avoid using them because of next task in this list.
- [ ] Add configuration and good defautls which does not change behavior for upgraded installations
- [ ] Please do not forget method docblocks

What do you think?

**Comment by MattBearson** [comment]
Thank you @stevenklar for review, I think you are totally right. This is a really good idea to move it into Consumer class and probably instead methods better to use values from `properties` in config\amqp.php.
What do you think?

**Comment by stevenklar** [comment]
Yep. Thats exactly what I thought. :-)

**Comment by MattBearson** [comment]
@stevenklar please, check it again, don't know, should I add a php doc description about `basic_qos`?

File: src/Amqp.php [hunk 1]
```diff
@@ -51,6 +50,8 @@ public function consume($queue, Closure $callback, $properties = [])
             ->mergeProperties($properties)
             ->setup();
 
+        $consumer->getChannel()->basic_qos($this->prefetch_size, $this->prefetch_count, $this->a_global);
```
Comment (stevenklar) [inline src/Amqp.php]: I guess this line could be deleted now.

File: src/Consumer.php [hunk 2]
```diff
@@ -45,6 +45,11 @@ public function consume($queue, Closure $closure)
 
             $object = $this;
 
+            $this->getChannel()->basic_qos(
+                $this->getProperty('qos_prefetch_size'),
+                $this->getProperty('qos_prefetch_count'),
+                $this->getProperty('qos_a_global'));
```
Comment (stevenklar) [inline src/Consumer.php]:
```suggestion
                $this->getProperty('qos_a_global')
            );
```

**Comment by MattBearson** [comment]
good catch, thanks, updated the branch

File: src/Amqp.php (hunk 1 above)
Comment (MattBearson) [reply src/Amqp.php]: fixed it

**Comment by stevenklar** [comment]
Please give me some time. I need to check if this feature need to be opt-in instead of being a default for every user.

**Comment by MattBearson** [comment]
yes, sure. I think we can add an additional parameter, like `qos_enable = false` by default. You are right, maybe not all users really need it.
Well about prefetch default value is 0, the same as `null` actually and it means "unlimited"

**Comment by stevenklar** [comment]
If you would like to add the configuration ```'qos' => false,``` with the conditional we can merge this and keep it deactivated for people don't using it.

Thanks!

**Comment by MattBearson** [comment]
yeah, I added it, this is really the best way to save backward compatibility and give possibility do not use it if don't need basic_qos at all

File: src/Consumer.php [hunk 3]
```diff
@@ -39,12 +39,22 @@ public function consume($queue, Closure $closure)
                 no_local: Don't receive messages published by this consumer.
                 no_ack: Tells the server if the consumer will acknowledge the messages.
                 exclusive: Request exclusive consumer access, meaning only this consumer can access the queue
-                nowait:
+                nowait: basic_qos() method is using to tell RabbitMQ not to give more than one message to a worker at a time.
```
Comment (stevenklar) [inline src/Consumer.php]:
```suggestion
                nowait:
                basic_qos: The consumer-prefetch make it possible to limit the number of unacknowledged messages on a channel (or connection) when consuming.
```
I might fix the explaination for nowait in a different PR.

**Comment by stevenklar** [comment]
Just correct the explanation and this get merged.

**Comment by MattBearson** [comment]
I see, yes, sure

**Pull request merged**